ambari-server/src/main/java/org/apache/ambari/server/checks/AmbariMetricsHadoopSinkVersionCompatibilityCheck.java

```diff
@@ -0,0 +1,215 @@
+    Map<String, Object> properties = new LinkedHashMap<>();
+    properties.put(RequestResourceProvider.REQUEST_CLUSTER_NAME_PROPERTY_ID, clusterName);
+
+    Set<Map<String, Object>> filterSet = new HashSet<>();
+    Map<String, Object> filterMap = new HashMap<>();
+    filterMap.put(RequestResourceProvider.SERVICE_ID, "AMBARI_METRICS");
+    filterMap.put(RequestResourceProvider.COMPONENT_ID, "METRICS_MONITOR");
+    filterMap.put(RequestResourceProvider.HOSTS_ID, StringUtils.join(hosts,","));
+    filterSet.add(filterMap);
+
+    properties.put(RequestResourceProvider.REQUEST_RESOURCE_FILTER_ID, filterSet);
+    propertiesSet.add(properties);
+
+    Map<String, String> requestInfoProperties = new HashMap<>();
+    requestInfoProperties.put(RequestResourceProvider.COMMAND_ID, "CHECK_HADOOP_SINK_VERSION");
+    requestInfoProperties.put(RequestResourceProvider.REQUEST_CONTEXT_ID, "Pre Upgrade check for compatible Hadoop Metric " +
+      "Sink version on all hosts.");
+
+    Request request = PropertyHelper.getCreateRequest(propertiesSet, requestInfoProperties);
+    try {
+      org.apache.ambari.server.controller.spi.RequestStatus response = provider.createResources(request);
+      Resource responseResource = response.getRequestResource();
+      String requestIdProp = PropertyHelper.getPropertyId("Requests", "id");
+      long requestId = (long) responseResource.getPropertyValue(requestIdProp);
+      LOG.debug("RequestId for AMS Hadoop Sink version compatibility pre check : " + requestId);
+
+      Thread.sleep(RETRY_INTERVAL);
+      PreUpgradeCheckStatus status;
+      int retry = 0;
+      LinkedHashSet<String> failedHosts = new LinkedHashSet<>();
+      while ((status = pollRequestStatus(requestId, failedHosts)).equals(PreUpgradeCheckStatus.RUNNING)
+        && retry++ < NUM_TRIES) {
+        if (retry != NUM_TRIES) {
+          Thread.sleep(RETRY_INTERVAL);
+        }
+      }
+
+      if (status.equals(PreUpgradeCheckStatus.SUCCESS)) {
+        prerequisiteCheck.setStatus(PrereqCheckStatus.PASS);
+      } else {
+        prerequisiteCheck.setStatus(PrereqCheckStatus.FAIL);
+        prerequisiteCheck.setFailReason(FAILED_REASON);
+        prerequisiteCheck.setFailedOn(failedHosts);
+      }
+    } catch (Exception e) {
+      LOG.error("Error running Pre Upgrade check for AMS Hadoop Sink compatibility. " + e);
+      prerequisiteCheck.setStatus(PrereqCheckStatus.FAIL);
+    }
+  }
+
+  /**
+   * Get the status of the requestId and also the set of failed hosts if any.
+   * @param requestId RequestId to track.
+   * @param failedHosts populate this argument for failed hosts.
+   * @return Status of the request.
+   * @throws Exception
+   */
+  private PreUpgradeCheckStatus pollRequestStatus(long requestId, Set<String> failedHosts) throws Exception {
+
+    List<RequestEntity> requestEntities = requestDAO.findByPks(Collections.singleton(requestId), true);
+    if (requestEntities != null && requestEntities.size() > 0) {
```

Comment:
Why do this if you're only using the first one?  Just call `requestDAO.findByPK()`

Reply:
This is because the findByPk has no option to refresh the JPA cache on find.

Reply:
This is an interesting issue. I'm OK with the workaround of forcing a refresh from the DB, but would there be another way to do this?